File: frontend/lib/geo.ts
// City coordinate lookup table (lat, lon) — covers major European + global metal cities
export const CITY_COORDS: Record<string, [number, number]> = {
  // Germany
  'berlin':      [52.5200, 13.4050],
  'hamburg':     [53.5511,  9.9937],
  'munich':      [48.1351, 11.5820],
  'münchen':     [48.1351, 11.5820],
  'cologne':     [50.9333,  6.9500],
  'köln':        [50.9333,  6.9500],
  'frankfurt':   [50.1109,  8.6821],
  'stuttgart':   [48.7758,  9.1829],
  'düsseldorf':  [51.2217,  6.7762],
  'dusseldorf':  [51.2217,  6.7762],
  'dortmund':    [51.5136,  7.4653],
  'essen':       [51.4556,  7.0116],
  'leipzig':     [51.3397, 12.3731],
  'bremen':      [53.0793,  8.8017],
  'dresden':     [51.0504, 13.7373],
  'hannover':    [52.3759,  9.7320],
  'nuremberg':   [49.4521, 11.0767],
  'nürnberg':    [49.4521, 11.0767],
  'mannheim':    [49.4875,  8.4660],
  'wacken':      [54.0783,  9.3740],
  // Austria
  'vienna':      [48.2082, 16.3738],
  'wien':        [48.2082, 16.3738],
  'graz':        [47.0707, 15.4395],
  'linz':        [48.3069, 14.2858],
  'salzburg':    [47.8095, 13.0550],
  'innsbruck':   [47.2692, 11.4041],
  // Switzerland
  'zurich':      [47.3769,  8.5417],
  'zürich':      [47.3769,  8.5417],
  'bern':        [46.9481,  7.4474],
  'basel':       [47.5596,  7.5886],
  'geneva':      [46.2044,  6.1432],
  // Scandinavia
  'oslo':        [59.9139, 10.7522],
  'stockholm':   [59.3293, 18.0686],
  'copenhagen':  [55.6761, 12.5683],
  'gothenburg':  [57.7089, 11.9746],
  'göteborg':    [57.7089, 11.9746],
  'helsinki':    [60.1699, 24.9384],
  'bergen':      [60.3913,  5.3221],
  'trondheim':   [63.4305, 10.3951],
  // UK & Ireland
  'london':      [51.5074, -0.1278],
  'manchester':  [53.4808, -2.2426],
  'birmingham':  [52.4862, -1.8904],
  'edinburgh':   [55.9533, -3.1883],
  'glasgow':     [55.8642, -4.2518],
  'dublin':      [53.3498, -6.2603],
  // Netherlands & Belgium
  'amsterdam':   [52.3676,  4.9041],
  'rotterdam':   [51.9244,  4.4777],
  'brussels':    [50.8503,  4.3517],
  'brussel':     [50.8503,  4.3517],
  'antwerp':     [51.2194,  4.4025],
  // France
  'paris':       [48.8566,  2.3522],
  'lyon':        [45.7640,  4.8357],
  'marseille':   [43.2965,  5.3698],
  'bordeaux':    [44.8378, -0.5792],
  'toulouse':    [43.6047,  1.4442],
  // Poland
  'warsaw':      [52.2297, 21.0122],
  'wroclaw':     [51.1079, 17.0385],
  'wrocław':     [51.1079, 17.0385],
  'krakow':      [50.0647, 19.9450],
  'kraków':      [50.0647, 19.9450],
  'gdansk':      [54.3520, 18.6466],
  'gdańsk':      [54.3520, 18.6466],
  'poznan':      [52.4064, 16.9252],
  // Czech Republic
  'prague':      [50.0755, 14.4378],
  'praha':       [50.0755, 14.4378],
  'brno':        [49.1951, 16.6068],
  // Italy
  'rome':        [41.9028, 12.4964],
  'milan':       [45.4642,  9.1900],
  'milano':      [45.4642,  9.1900],
  'florence':    [43.7696, 11.2558],
  'bologna':     [44.4949, 11.3426],
  'turin':       [45.0703,  7.6869],
  // Spain
  'madrid':      [40.4168, -3.7038],
  'barcelona':   [41.3851,  2.1734],
  'bilbao':      [43.2630, -2.9340],
  // USA
  'new york':    [40.7128, -74.0060],
  'los angeles': [34.0522,-118.2437],
  'chicago':     [41.8781, -87.6298],
  'portland':    [45.5231,-122.6765],
  'seattle':     [47.6062,-122.3321],
  'san francisco':[37.7749,-122.4194],
  'atlanta':     [33.7490, -84.3880],
  'denver':      [39.7392,-104.9903],
  'boston':      [42.3601, -71.0589],
  // Canada
  'toronto':     [43.6532, -79.3832],
  'montreal':    [45.5017, -73.5673],
  'vancouver':   [49.2827,-123.1207],
  // Other
  'tokyo':       [35.6762, 139.6503],
  'sydney':      [-33.8688, 151.2093],
  'melbourne':   [-37.8136, 144.9631],
}

export function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371
  const dLat = (lat2 - lat1) * Math.PI / 180
  const dLon = (lon2 - lon1) * Math.PI / 180
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
    Math.sin(dLon / 2) ** 2
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
}

export function distanceBetweenCities(city1: string, city2: string): number | null {
  const c1 = CITY_COORDS[city1.toLowerCase().trim()]
  const c2 = CITY_COORDS[city2.toLowerCase().trim()]
  if (!c1 || !c2) return null
  return Math.round(haversineKm(c1[0], c1[1], c2[0], c2[1]))
}

export function formatDistance(km: number): string {
  if (km < 1) return '< 1 km'
  return `${km} km`
}

File: backend/app/utils/geo.py
"""Shared geo utilities: city-to-coordinate lookup and Haversine distance."""
import math
from typing import Dict, Optional, Tuple

CITY_COORDS: Dict[str, Tuple[float, float]] = {
    # Germany
    'berlin': (52.520, 13.405), 'hamburg': (53.550, 9.994),
    'munich': (48.135, 11.582), 'münchen': (48.135, 11.582),
    'cologne': (50.933, 6.950), 'köln': (50.933, 6.950),
    'frankfurt': (50.111, 8.682), 'frankfurt am main': (50.111, 8.682),
    'düsseldorf': (51.225, 6.782), 'stuttgart': (48.776, 9.183),
    'leipzig': (51.340, 12.373), 'dresden': (51.050, 13.737),
    'bremen': (53.075, 8.808), 'hannover': (52.373, 9.738),
    'nuremberg': (49.453, 11.077), 'nürnberg': (49.453, 11.077),
    'wacken': (54.078, 9.374), 'dortmund': (51.514, 7.468),
    'bochum': (51.482, 7.217), 'essen': (51.457, 7.012),
    # Austria / Switzerland
    'vienna': (48.208, 16.374), 'wien': (48.208, 16.374),
    'graz': (47.071, 15.440), 'salzburg': (47.810, 13.055),
    'linz': (48.306, 14.286),
    'zurich': (47.377, 8.542), 'zürich': (47.377, 8.542),
    'bern': (46.948, 7.447), 'geneva': (46.204, 6.143), 'basel': (47.560, 7.590),
    # Nordics
    'oslo': (59.914, 10.752), 'bergen': (60.392, 5.324),
    'stockholm': (59.329, 18.069), 'gothenburg': (57.707, 11.967),
    'copenhagen': (55.676, 12.568), 'aarhus': (56.157, 10.211),
    'helsinki': (60.170, 24.938), 'tampere': (61.498, 23.761),
    'reykjavik': (64.136, -21.895),
    # UK / Ireland
    'london': (51.507, -0.128), 'manchester': (53.481, -2.243),
    'glasgow': (55.864, -4.252), 'edinburgh': (55.953, -3.188),
    'birmingham': (52.480, -1.903), 'leeds': (53.800, -1.549),
    'bristol': (51.455, -2.588), 'sheffield': (53.383, -1.466),
    'dublin': (53.350, -6.260), 'belfast': (54.597, -5.930),
    # Benelux
    'amsterdam': (52.368, 4.904), 'rotterdam': (51.924, 4.478),
    'the hague': (52.070, 4.300), 'utrecht': (52.090, 5.121),
    'brussels': (50.850, 4.352), 'antwerp': (51.221, 4.400),
    'luxembourg': (49.611, 6.132),
    # France
    'paris': (48.857, 2.352), 'lyon': (45.764, 4.836),
    'marseille': (43.297, 5.381), 'toulouse': (43.604, 1.444),
    'bordeaux': (44.837, -0.579), 'lille': (50.629, 3.057),
    # Iberia
    'madrid': (40.417, -3.704), 'barcelona': (41.385, 2.173),
    'seville': (37.388, -5.982), 'bilbao': (43.262, -2.925),
    'lisbon': (38.717, -9.139), 'porto': (41.157, -8.629),
    # Italy
    'rome': (41.903, 12.496), 'milan': (45.464, 9.190),
    'naples': (40.851, 14.268), 'turin': (45.070, 7.687),
    'bologna': (44.494, 11.343), 'florence': (43.769, 11.256),
    # Eastern Europe
    'warsaw': (52.230, 21.012), 'krakow': (50.061, 19.937),
    'prague': (50.076, 14.438), 'brno': (49.196, 16.607),
    'budapest': (47.498, 19.040), 'bucharest': (44.426, 26.103),
    'sofia': (42.698, 23.322), 'belgrade': (44.787, 20.457),
    'zagreb': (45.815, 15.982), 'ljubljana': (46.051, 14.506),
    'bratislava': (48.148, 17.107), 'tallinn': (59.437, 24.754),
    'riga': (56.946, 24.106), 'vilnius': (54.687, 25.280),
    # Greece / Turkey
    'athens': (37.984, 23.728), 'thessaloniki': (40.636, 22.945),
    'istanbul': (41.015, 28.979), 'ankara': (39.920, 32.854),
    # North America
    'new york': (40.713, -74.006), 'los angeles': (34.052, -118.244),
    'chicago': (41.878, -87.630), 'toronto': (43.653, -79.383),
    'montreal': (45.501, -73.567), 'vancouver': (49.283, -123.121),
    'san francisco': (37.775, -122.419), 'seattle': (47.606, -122.332),
    'boston': (42.360, -71.059), 'dallas': (32.776, -96.797),
    'houston': (29.760, -95.370), 'atlanta': (33.749, -84.388),
    'miami': (25.775, -80.208), 'phoenix': (33.448, -112.074),
    'denver': (39.739, -104.984), 'detroit': (42.332, -83.046),
    'portland': (45.523, -122.676), 'las vegas': (36.175, -115.137),
    'minneapolis': (44.977, -93.265), 'nashville': (36.174, -86.768),
    # Oceania
    'sydney': (-33.869, 151.209), 'melbourne': (-37.814, 144.963),
    'brisbane': (-27.468, 153.028), 'perth': (-31.952, 115.861),
    'auckland': (-36.852, 174.763),
    # Asia
    'tokyo': (35.676, 139.650), 'osaka': (34.693, 135.502),
    'seoul': (37.566, 126.978), 'beijing': (39.906, 116.391),
    'shanghai': (31.228, 121.474), 'hong kong': (22.396, 114.109),
    'singapore': (1.352, 103.820), 'taipei': (25.047, 121.519),
    # South America
    'buenos aires': (-34.603, -58.382), 'sao paulo': (-23.549, -46.633),
    'santiago': (-33.459, -70.648), 'bogota': (4.711, -74.072),
    # Russia
    'moscow': (55.751, 37.617), 'saint petersburg': (59.939, 30.316),
}


def _haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    R = 6371.0
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (math.sin(dlat / 2) ** 2
         + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2))
         * math.sin(dlon / 2) ** 2)
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def resolve_coords(city: Optional[str]) -> Optional[Dict[str, float]]:
    """Return {'lat': ..., 'lon': ...} for a known city name, or None."""
    if not city:
        return None
    coords = CITY_COORDS.get(city.lower().strip())
    if coords is None:
        return None
    return {"lat": coords[0], "lon": coords[1]}


def distance_between_cities_km(
    city1: Optional[str], city2: Optional[str]
) -> Optional[float]:
    """Haversine distance in km between two named cities, or None if unknown."""
    c1 = resolve_coords(city1)
    c2 = resolve_coords(city2)
    if c1 is None or c2 is None:
        return None
    return _haversine_km(c1["lat"], c1["lon"], c2["lat"], c2["lon"])
